internal/action/action_impl.go

```diff
@@ -0,0 +1,197 @@
+package action
+
+import (
+	"fmt"
+	"math"
+	"os"
+	"strings"
+
+	"github.com/modern-magic/grm/internal"
+	"github.com/modern-magic/grm/internal/fs"
+	"github.com/modern-magic/grm/internal/logger"
+	"github.com/modern-magic/grm/internal/registry"
+)
+
+type actionImpl struct {
+	fs         fs.FS
+	source     map[string]registry.RegsitryInfo
+	userSource map[string]registry.RegsitryInfo
+	args       []string
+}
+
+type ActionOptions struct {
+	Source     map[string]registry.RegsitryInfo
+	UserSource map[string]registry.RegsitryInfo
+	Args       []string
+}
+
+func NewAction(options ActionOptions) Action {
+	return &actionImpl{
+		fs:         fs.NewFS(),
+		source:     options.Source,
+		userSource: options.UserSource,
+		args:       options.Args,
+	}
+}
+
+func loadRegistryFromSource(source map[string]registry.RegsitryInfo, args []string, invork func(r *RegistryDataSource) int) int {
+	defer func() {
+		if err := recover(); err != nil {
+			fmt.Println(err)
+			return
+		}
+	}()
```

Comment:
Why use "defer" here？Is it because this is the entry point for all operations?

Reply:
> Is it because this is the entry point for all operations?

Because the function `loadRegistryFromSource` will get the alias from `os.Args`. But we know if no value entered. It will cause out of bounds.So i should catch the error and exit the process.( I think we don't handle the error can cause a bad DX.

Reply:
Can we provide a global recover, such as the main function to call `defer func() { if err := recover ...}`, so that if a panic occurs before loadRegistryFromSource, we can catch the error.

Reply:
I think there should be a corresponding recover to catch the error when each goroutine is created. This might be a more appropriate way to catch panic. Of course, if you have other ideas, feel free to discuss them!

Reply:
My misstake. I didn't created new goroutine for each command ( Test is Exception. So cature the panic in main theads is right.

Reply:
And just a suggestion. In other language like `javascript`. we use `try catch` block syntax to catch error. But in golang shall we need a glob capture?

Reply:
这样其实也挺好，只是全局 recover 是我个人爱好😂，不喜欢去针对一个函数调用 panic 和 recover。